### DISCUSSION TITLE
Support JSX syntax for Astro components

### DISCUSSION THREAD

**Discussion by mb21** [withastro/roadmap#716] [Proposal]
# Summary

Support for .jsx/.tsx syntax in addition to (or replacing) the syntax found in .astro files.

# Background & Motivation

Sorry if this has been considered before.

We've been working with Astro full-time for a few weeks now, and quickly ran into the following issues:

- https://github.com/withastro/language-tools/issues/476 (which Solid Start solves by early returning a [`<Navigation />` component](https://start.solidjs.com/core-concepts/routing#redirecting))
- A rather big number of [compiler bugs](https://github.com/withastro/compiler/issues?q=is%3Aissue+label%3A%22-+P4%3A+important%22%2C%22-+P5%3A+urgent%22)

There is also these discussions that would profit from JSX:

- https://github.com/withastro/roadmap/discussions/833
- https://github.com/withastro/roadmap/issues/533
- https://github.com/withastro/roadmap/discussions/723
- https://github.com/withastro/roadmap/discussions/168

I remember Ryan Carniato (SolidJS developer) saying in a video that he's seen other framework's pain with rolling their own syntax/parsers (I now have a suspicion that he was referring to Astro), and that's why [Solid went with JSX](https://www.solidjs.com/tutorial/introduction_jsx).

Yes, I know the [JSX standard](https://facebook.github.io/jsx/) has stagnated, and no I don't think Astro would ~need to follow the standard 100%~ have to follow all the same conventions as React. Solid doesn't either. For example Solid still uses `class` instead of `className` and Astro could do the same.

I realize there are some details (like slots) to figure out how to translate to JSX, but these seem all solvable.

Finally, I agree the no-boilerplate feel of just putting HTML into a file and having it working without any JS function declarations and exports is appealing at first sight and to newcomers that don't know JS very well. But it seems to me in actual production usage, there are very few components that don't take any props and don't have any frontmatter at all. And for those, you could just use an `.html` file. And in both, Astro templates and JSX, we’re using a template system that’s embedded in its host language; meaning you can use arbitrary JavaScript syntax within the template. So I feel like the use-case for people not knowing any JavaScript at all is not really worth optimizing for?

Another example is that to pass in children to components, in Astro you have to use a [slot](https://docs.astro.build/en/basics/astro-components/#slots), similar to [HTML’s slot element](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/slot), somehow mimicking how a poly-fill for native web functionality would work. But we’re running on a server here and rendering HTML to a string that will be sent over the network. There is no DOM, and Web Components will never be implemented in Node/Deno/whatever. In React/Solid, child elements are simply passed as `props.children` – just like any other component prop.

So all in all, relying on JavaScript/JSX would require the developer to learn fewer different concepts than the weird mix of HTML-related concepts and JS that current astro components are. Finally, Astro could continue to support the current .astro format.

# Goals

- Reduce Astro implementation/maintenance overhead for custom syntax
- Reduce bugs in template parsing
- Users would avoid having to install a separate VScode plugin and it would be easier to work in IDEs that currently don't have an Astro plugin (like [Zed](https://github.com/zed-industries/zed))
- A lot of users are already very familiar with JSX, so they wouldn't have to learn where the current Astro syntax slightly differs from JSX (because it is already very similar to JSX.)

# Example

No need for an example. You all know what JSX looks like, that's the point ;-)

**Comment by Princesseuh** [comment]
> Yes, I know the [JSX standard](https://facebook.github.io/jsx/) has stagnated, and no I don't think Astro would need to follow the standard 100% (Solid doesn't either). For example Solid still uses class instead of className and Astro could do the same.

This is not JSX, this is React. Solid 100% follow JSX, otherwise it wouldn't be compatible with Prettier, TypeScript etc etc. Astro would have to do the same, otherwise we'd run into similar issues.

See this tweet by Ryan explaining why things like `<div {something}>` doesn't exist in Solid: https://twitter.com/RyanCarniato/status/1545476593207783424

**Comment by wakaztahir** [comment]
I have a similar proposal, That is to allow using `client:load` `client:only` directives inside framework components

My initial thought is that Astro team would make libraries for `React` & `Solid` which would provide directive prop or something similar that I would be able to include in my Solid UI library for example and I would be able to tell Astro which components should be rendered on server and which are client only, I could do it in comments too

```typescript
// @astro:client:only
export function MySolidUIComp() {
  return (<></>)
}
```

Sometimes whole app has state which is very complex and `createSignal` and `useState` are very handy

This would give complete control over how components are rendered using Astro

Then when I consume the library I can use components of that library in my project, I can write whole pages in my framework language without having to separate components to be able to have access to different islands

This would aid in UI libraries, UI library components would provide components that tell already which components require JavaScript, simplifying the process for the user who consumes the library

**Comment by fweth** [comment]
Any updates on this? Alternatively, it would be great to support frontmatter in React or Preact components. For example, by exporting `getStaticProps` or something similar. This would also eliminate the current unintuitive behavior where frontmatter data can't be used in `getStaticPaths`. That behavior makes sense logically, but it doesn't align with typical JavaScript logic.

By allowing `getStaticPaths` and `getStaticProps` to be exported as separate functions, the distinction would be much clearer.

**Comment by andersk** [comment]
The current `.astro` syntax is really frustrating due to an apparently unfixable issue in prettier-plugin-astro:

- withastro/prettier-plugin-astro#308
- https://github.com/withastro/astro/issues/12554#issuecomment-3141478291

Maybe this sounds nit-picky, but it’s actually a big problem. Automatic formatting is essential in large projects with many developers. But when it wrongly inserts or removes significant whitespace, that often breaks the intended display of the page. There doesn’t seem to be a reasonable workaround.

(I understand that the motivation for Astro’s whitespace rules was to be “closer to HTML’s”—but that’s not helpful if tools like formatters don’t understand that.)

If we had an option to use real JSX for Astro components, or at least JSX-compatible whitespace handling within embedded JSX expressions, we could sidestep this problem.

Edit: Cool, JSX-compatible whitespace is now the default with Astro 7 (withastro/astro#16965).

**Reply by Princesseuh** [reply]
As I wrote in one of these issues, it's fixable (and much easier than making Astro support JSX components), it's just that I (me) don't have the time for it and no one else has tried to contribute a fix despite me offering mentoring many times.

(There's also the workaround of using markup_fmt instead of prettier-plugin-astro, that being a reasonable workaround is up to you, though.)

**Reply by fweth** [reply]
I'd be curious, do developers prefer HTML or JSX whitespace behavior? HTML whitespace has [problems](https://blog.dwac.dev/posts/html-whitespace/), but does JSX fix them?
